### PR TITLE
Make local repository consumption more "user friendly"

### PR DESCRIPTION
**Pull request by superm1**
Currently local repositories require that you:
1. Put files into the correct place in filesystem
2. Clone lvfs-website.
3. Install any missing dependencies.
4. Run `./local.py` with proper arguments to generate metadata

With this change metadata is not needed for local repositories.  Instead set a directory in the conf file and then drop CAB files in the directory.  It will be parsed when daemon starts up and local metadata generated.

Since the metadata is generated from CAB files, it's no longer a true statement that the requirements check for a URI is true.
  
Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/hughsie/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

File: src/fu-common-cab.c [hunk 1]
```diff
@@ -187,6 +187,8 @@ fu_common_store_from_cab_file (XbBuilder *builder, GCabCabinet *cabinet,
 	g_autoptr(GError) error_local = NULL;
 	g_autoptr(XbBuilderSource) source = xb_builder_source_new ();
 
+	/* rewrite to be under a components root */
+	xb_builder_source_set_prefix (source, "components");
```
Comment (hughsie) [inline src/fu-common-cab.c]: I can't believe I didn't think of that. It's exactly what the `xb_builder_source_set_prefix` call is useful for. Maybe add one empty line under this between the other comment, but I'm really nitpicking here.

File: src/fu-engine.c [hunk 2]
```diff
@@ -1809,6 +1809,65 @@ fu_engine_appstream_upgrade_cb (XbBuilderFixup *self,
...
+		g_autoptr(GBytes) blob = NULL;
+		g_autoptr(XbSilo) silo = NULL;
+		g_autoptr(XbBuilderSource) source = xb_builder_source_new ();
+		g_autoptr(XbBuilderNode) custom = NULL;
+		g_autofree gchar *xml = NULL;
+
+		const gchar *fn = g_ptr_array_index (files, i);
+		if (fnmatch (pattern, fn, 0) != 0) {
+			continue;
+		}
```
Comment (hughsie) [inline src/fu-engine.c]: No braces pls.

**Comment by hughsie** [comment]
Code nits aside, I have two reservations:

 * If I dump 100 cab files in that directory, how long will it take to start up fwupdtool? I think most of the cab files are going to be uncompressed, but that's still lots of i/o just to read off disk and then parse a very small file buried deep inside. I do wonder whether fwupd should go to the trouble of writing out a metadata file for future loads, although it would have to be somewhere not `/usr/share/...` -- but see next point.

* Some non-public deployments of fwupd don't use the LVFS directly (no Internet access), and instead use a `LOCAL` metadata store where the metadata is stored alongside the majority of the LVFS firmware on a NFS share. I really don't think we want to be parsing all the .cab files over NFS at fwupd startup.

I think leaving `FWUPD_REMOTE_KIND_LOCAL` as _load metadata xml from local storage_ is the best plan, and then _adding_ something like `FWUPD_REMOTE_KIND_LOCAL_AUTO`, `FWUPD_REMOTE_KIND_LOCAL_ARCHIVES` or `FWUPD_REMOTE_KIND_DIRECTORY` which does the auto-loading-cab-file thing you're proposing here.

File: src/fu-common-cab.c (hunk 1 above)
Comment (superm1) [reply src/fu-common-cab.c]: Yeah it had me scratching my head too, it looked like it fit the bill perfectly.

**Comment by superm1** [comment]
OK, so I've added a new type for "directory" and a little documentation on how exactly to use it.  It was working yesterday, but functionally I appear to have broken something in the process.
Let me know what you think of this now and I'll figure out what I broke.

In terms of startup time going high because of this, yes I think it's a valid concern for some use cases.  The case I'm thinking through is <10 firmware items built into an image and you scan those when you start the tool or daemon.  I don't think it makes a lot of sense to set up caching for that small of a number.  If it's 100 then yes it does need that.

**Comment by hughsie** [comment]
Looks pretty good. More fixups, but basically what I imagined.

**Comment by superm1** [comment]
It seems the other night my test wasn't fully valid, it was getting some metadata (particularly checksums) from an LVFS source even if it was getting the file from the local one.  From what I can tell the remaining problem that in `fu-common-cab.c` if no checksums are found on the CAB they don't get calculated and inserted into the metadata at that time.
This is problematic as at least the CAB I grabbed from LVFS for these tests didn't have that set.

Ignoring the checksums (as in how ab5e5dc4fdc171939de7a142c5695bdd769d084f currently does) leads to a problem trying to match up how the the release URI to how the local sources is supposed to work in `fu-tool.c` and `fu-util.c`.  It didn't look obvious to me, is there an easy way to inject that metadata at the right time?  It seems like it should be plopped in during `fu_common_store_from_cab_release` but i'm not really sure how to make a new node at that time.  So maybe it needs to happen earlier.  If you've got a good way you would like to see this happen please feel free to commit it to this branch.

**Comment by hughsie** [comment]
I think we need the checksums. I'll have a play in your branch, give me a couple of hours.

**Comment by hughsie** [comment]
Okay, I think it works. The two other related PRs needs to be merged before this one, and then if it works can you squash it down, test everything and we'll do one last quick sanity check before merging.

File: src/fu-engine.c [hunk 3]
```diff
@@ -1815,6 +1821,86 @@ fu_engine_appstream_upgrade_cb (XbBuilderFixup *self,
...
+	/* add each source */
+	pattern = g_build_filename (path, "*.cab", NULL);
+	for (guint i = 0; i < files->len; i++) {
+		g_autoptr(XbBuilderNode) custom = NULL;
+		g_autoptr(XbBuilderSource) source = NULL;
+		g_autoptr(GError) error_local = NULL;
+		const gchar *fn = g_ptr_array_index (files, i);
+
+		/* check is cab file */
+		if (fnmatch (pattern, fn, 0) != 0) {
```
Comment (hughsie) [inline src/fu-engine.c]: Was there a reason to use `fnmatch` rather than just `g_str_has_suffix(fn, ".cab")`?

File: src/fu-engine.c [hunk 4]
```diff
@@ -2523,10 +2621,10 @@ fu_engine_add_releases_for_device_component (FuEngine *self,
 		/* fall back to quirk-provided value */
 		if (fwupd_release_get_install_duration (rel) == 0)
 			fwupd_release_set_install_duration (rel, fu_device_get_install_duration (device));
-
```
Comment (hughsie) [inline src/fu-engine.c]: I think this blank line should stay...

File: src/fu-util.c [hunk 5]
```diff
@@ -1973,17 +1973,19 @@ fu_util_update_device_with_release (FuUtilPrivate *priv,
...
 			const gchar *fn_cache = fwupd_remote_get_filename_cache (remote);
 			g_autofree gchar *path = g_path_get_dirname (fn_cache);
 
-			/* install with flags chosen by the user */
 			fn = g_build_filename (path, uri_tmp, NULL);
+		} else if (fwupd_remote_get_kind (remote) == FWUPD_REMOTE_KIND_DIRECTORY) {
+			fn = g_strdup (uri_tmp + 7);
+		}
+		/* install with flags chosen by the user */
+		if (fn != NULL)
```
Comment (hughsie) [inline src/fu-util.c]: why would fn be NULL? Isn't that an error we need to handle?

Comment (hughsie) [reply src/fu-util.c]: Ohh, ignore me, I see now. I think adding back the braces would make this clearer.

File: src/fu-engine.c [hunk 6]
```diff
@@ -304,6 +304,12 @@ fu_engine_set_release_from_appstream (FuEngine *self,
 		if (uri == NULL)
 			uri = g_strdup (tmp);
 		fwupd_release_set_uri (rel, uri);
+	} else if (remote != NULL &&
+		   fwupd_remote_get_kind (remote) == FWUPD_REMOTE_KIND_DIRECTORY) {
+		g_autofree gchar *uri = NULL;
+		tmp = xb_node_query_text (component, "../custom/value[@key='fwupd::FilenameCache']", NULL);
```
Comment (hughsie) [inline src/fu-engine.c]: Should we test for tmp==NULL and return an error in that case?

File: data/remotes.d/automatic-vendor.conf [hunk 7]
```diff
@@ -0,0 +1,7 @@
+[fwupd Remote]
+# this configuration provides dynamically generated metadata shipped by the OS vendor and can
+# be found in @datadir@/fwupd/remotes.d/vendor/firmware
+Enabled=false
+Title=Vendor (Automatic)
```
Comment (hughsie) [inline data/remotes.d/automatic-vendor.conf]: I think a new .conf file makes sense, but I'm not sure about the `automatic-vendor.conf` name. Maybe `vendor-directory.conf` or something?

File: data/remotes.d/README.md [hunk 8]
```diff
@@ -39,6 +39,20 @@ certificate. If this is the case also change `Keyring=gpg` or `Keyring=pkcs7`
...
 
+Automatic metadata generation
+=============================
+`fwupd` and `fwupdtool` support automatically generating metadata for a remote
+by configuring it to be a *directory* type. This is very convenient if you want to dynamically add firmware from multiple packages while generating the image but there are a few deficiencies:
+* There will be a performance impact of starting the daemon or tool measured by O(# CAB files)
+* It's not possible to verify metadata signature and any file validation should be part of the image validation.
+
+To enable this:
+* Change `/etc/fwupd/remotes.d/vendor.conf` to have `Enabled=true`
```
Comment (hughsie) [inline data/remotes.d/README.md]: You mean automatic-vendor.conf, no?

File: src/fu-engine.c (hunk 6 above)
Comment (superm1) [reply src/fu-engine.c]: it's a void function, but I'll add a check for `tmp != NULL`

File: data/remotes.d/README.md (hunk 8 above)
Comment (superm1) [reply data/remotes.d/README.md]: Yes, thanks fixed

File: data/remotes.d/automatic-vendor.conf (hunk 7 above)
Comment (superm1) [reply data/remotes.d/automatic-vendor.conf]: OK, renamed to `vendor-directory.conf`

File: src/fu-util.c (hunk 5 above)
Comment (superm1) [reply src/fu-util.c]: Added braces

**Pull request merged**